core/src/test/java/org/apache/iceberg/TestInternalData.java

```diff
@@ -0,0 +1,245 @@
+        InternalData.write(fileFormat, outputFile).schema(simpleSchema()).build()) {
+      appender.addAll(testData);
+    }
+
+    InputFile inputFile = fileIO.newInputFile(outputFile.location());
+    List<PartitionData> readRecords = Lists.newArrayList();
+
+    try (CloseableIterable<PartitionData> reader =
+        InternalData.read(fileFormat, inputFile)
+            .project(simpleSchema())
+            .setRootType(PartitionData.class)
+            .build()) {
+      for (PartitionData record : reader) {
+        readRecords.add(record);
+      }
+    }
+
+    assertThat(readRecords).hasSize(testData.size());
+
+    for (int i = 0; i < testData.size(); i++) {
+      Record expected = testData.get(i);
+      PartitionData actual = readRecords.get(i);
+
+      assertThat(actual.get(0, Long.class)).isEqualTo(expected.get(0, Long.class));
+      assertThat(actual.get(1, String.class)).isEqualTo(expected.get(1, String.class));
+    }
+  }
+
+  @TestTemplate
+  public void testCustomTypeForNestedField() throws IOException {
+    FileFormat fileFormat = FileFormat.fromString(format);
+    OutputFile outputFile = fileIO.newOutputFile(tempDir.resolve("test." + format).toString());
+
+    List<Record> testData = createNestedTestRecords();
+
+    try (FileAppender<Record> appender =
+        InternalData.write(fileFormat, outputFile).schema(NESTED_SCHEMA).build()) {
+      appender.addAll(testData);
+    }
+
+    InputFile inputFile = fileIO.newInputFile(outputFile.location());
+    List<Record> readRecords = Lists.newArrayList();
+
+    try (CloseableIterable<Record> reader =
+        InternalData.read(fileFormat, inputFile)
+            .project(NESTED_SCHEMA)
+            .setCustomType(2, TestCustomRow.class)
+            .build()) {
+      for (Record record : reader) {
+        readRecords.add(record);
+      }
+    }
+
+    assertThat(readRecords).hasSize(testData.size());
+
+    for (int i = 0; i < testData.size(); i++) {
+      Record expected = testData.get(i);
+      Record actual = readRecords.get(i);
+
+      assertThat(actual.get(0, Long.class)).isEqualTo(expected.get(0, Long.class));
+
+      Object expectedNested = expected.get(1);
+      Object actualNested = actual.get(1);
+
+      if (expectedNested == null && actualNested == null) {
+        continue;
+      }
+
+      if (actualNested != null) {
+        assertThat(actualNested)
+            .as("Custom type should be TestCustomRow, but was: " + actualNested.getClass())
+            .isInstanceOf(TestCustomRow.class);
+        TestCustomRow customRow = (TestCustomRow) actualNested;
+        Record expectedRecord = (Record) expectedNested;
+
+        assertThat(customRow.get(0, Long.class))
+            .isEqualTo(expectedRecord.get(0, Long.class)); // inner_id
+        assertThat(customRow.get(1, String.class))
+            .isEqualTo(expectedRecord.get(1, String.class)); // inner_name
+      }
+    }
+  }
+
+  private List<Record> createSimpleTestRecords() {
```

Comment:
instead of the two create methods, maybe use `RandomInternalData.generate(Schema schema, int numRecords, long seed)` ?

Reply:
That's fair, the only important thing is having a few nulls